core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -232,6 +233,16 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
         if (canExecuteScript(lang, scriptEngineService, script.getType(), scriptContext) == false) {
             throw new ScriptException("scripts of type [" + script.getType() + "], operation [" + scriptContext.getKey() + "] and lang [" + lang + "] are disabled");
         }
+
+        // special exception to prevent expressions from compiling as update or mapping scripts
+        boolean expression = scriptEngineService instanceof ExpressionScriptEngineService;
```

Comment:
I consider adding getSupportedScriptContexts() to the ScriptEngineService a breaking change because it will force all existing ScriptEngines including plugins outside of our control to add a method since ScriptEngineService is an interface.  It's also tricky because of the global settings and granular settings for each ScriptContext that's allowed need to be thought about.  For global settings, I guess we just ignore them and say that granular settings override them, but for granular settings, if I set to expressions to be allowed to run with update or mapping in the settings that needs to throw an error.  It also affects the behavior of some of the tests that are randomized.  I do agree after looking at the code again that I overstated how large the change is, but I do think it may be tricky to get the settings and tests working correctly.  Anyway, I am going to push this to another PR for now.